src/olympia/amo/middleware.py

```diff
@@ -188,15 +191,50 @@ def process_request(self, request):
 
 
 class ReadOnlyMiddleware(object):
+    """Middleware that announces a downtime which for us usually means
+    putting the site into read only mode.
+
+    Supports issuing `Retry-After` header.
+    """
+    ERROR_MSG = _(
+        u'Some features are temporarily disabled while we '
+        u'perform website maintenance. We\'ll be back to '
+        u'full capacity shortly.')
+
+    API_HEADER_NAME = 'X-AMO-Read-Only'
+
+    def _render_api_error(self):
+        response = JsonResponse({'error': self.ERROR_MSG}, status=503)
+        response[self.API_HEADER_NAME] = 'true'
+        if settings.READ_ONLY_RETRY_AFTER is not None:
+            response['Retry-After'] = settings.READ_ONLY_RETRY_AFTER.seconds
+        return response
 
     def process_request(self, request):
-        if request.method == 'POST':
+        if not settings.READ_ONLY:
+            return
+
+        if request.path.startswith('/api/'):
```

Comment:
random observation: maybe we need to set something on the request in one of the middlewares to indicate it's an API request once so we're not testing the path in different places. Not in scope for this patch, just sayin'

Reply:
Agreed! Filed https://github.com/mozilla/addons-server/issues/9170 about it